dace/transformation/dataflow/redundant_array.py

```diff
@@ -154,28 +173,46 @@ def gnode(nname):
         in_desc = sdfg.arrays[in_array.data]
         out_desc = sdfg.arrays[out_array.data]
 
-        # If arrays are not of the same shape, modify first array to reshape
-        # instead of removing it
-        if len(in_desc.shape) != len(out_desc.shape) or any(
-                i != o for i, o in zip(in_desc.shape, out_desc.shape)):
+        # 1. Get edge e1 and extract subsets for arrays A and B
+        e1 = graph.edges_between(in_array, out_array)[0]
+        a1_subset, b_subset = _validate_subsets(e1, sdfg.arrays)
+
+        # If the memlet does not cover the remove array, create a view.
+        if any(m != a for m, a in zip(a1_subset.size(), in_desc.shape)):
             sdfg.arrays[in_array.data] = data.View(
                 in_desc.dtype, in_desc.shape, True, in_desc.allow_conflicts,
                 out_desc.storage, out_desc.location, in_desc.strides,
                 in_desc.offset, out_desc.may_alias,
                 dtypes.AllocationLifetime.Scope, in_desc.alignment,
                 in_desc.debuginfo, in_desc.total_size)
             return
-
-        for e in graph.in_edges(in_array):
-            # Modify all incoming edges to point to out_array
-            path = graph.memlet_path(e)
-            for pe in path:
-                if pe.data.data == in_array.data:
-                    pe.data.data = out_array.data
-
-            # Redirect edge to out_array
-            graph.remove_edge(e)
-            graph.add_edge(e.src, e.src_conn, out_array, e.dst_conn, e.data)
+        
+        # 2. Iterate over the e2 edges and traverse the memlet tree
+        for e2 in graph.in_edges(in_array):
+            path = graph.memlet_tree(e2)
+            for e3 in path:
+                # 2-a. Extract subsets for array B and others
+                other_subset, a3_subset = _validate_subsets(
+                    e3, sdfg.arrays, dst_name=in_array.data)
+                # 2-b. Modify memlet to match array B.
+                e3.data.data = out_array.data
+                a3_subset.offset(a1_subset, negative=True)
+                if isinstance(b_subset, subsets.Indices):
+                    tmp = copy.deepcopy(b_subset)
+                    tmp.offset(a3_subset, negative=False)
```

Comment:
you can use `offset_new` for this, which creates an out-of-place (new) offsetted subset